Implement string.concat() and bytes.concat()

seanyoung opened the pull request:
This is what solc implements. Solang has implement a + b for string concatenation, which this PR removes.

Fixes: https://github.com/hyperledger/solang/issues/1558

LucasSte commented:
> This is what solc implements. Solang has implement a + b for string concatenation, which this PR removes.
> 
> Fixes: #1558

Removing `a+b` is a breaking change and render existing contracts incompatible. Shouldn't we first warn that `a+b` is deprecated and then remove it in a later release?

seanyoung commented on stdlib/stdlib.c:
```diff
@@ -52,6 +52,8 @@ void __memcpy(void *_dest, const void *_src, uint32_t length)
     {
         *dest++ = *src++;
     }
+
+    return dest;
```
It makes concatenation easier in emit; the code can simply use the return value for the destination of the next part that needs to be concatenated.

LucasSte replied:
Why is it easier? Before you only need to construct a function call inside a for loop without changing the first argument, now you need a return that is exactly the same value as the argument.

seanyoung replied:
It does not return the same value as its argument. It returns the dest + len

seanyoung commented on tests/polkadot_tests/strings.rs:
```diff
@@ -144,7 +144,12 @@ fn string_concat() {
         contract foo {
             function test() public {
-                assert(hex"41424344" == "AB" + "CD");
+                assert(hex"41424344" == bytes.concat("AB", "CD", bytes(string.concat())));
```
There is a test with three arguments in the `string_abi_decode()` test. I've added some more tests.

seanyoung commented on src/codegen/events/polkadot.rs:
```diff
@@ -124,13 +124,23 @@ impl EventEmitter for PolkadotEventEmitter<'_> {
                 ty: Type::DynamicBytes,
-                left: prefix,
-                right: value,
+                size: Expression::NumberLiteral {
+                    loc,
+                    ty: Type::Uint(32),
+                    value: first_prefix.len().into(),
+                }
+                .into(),
+                initializer: Some(first_prefix),
+            };
```
Fixed

seanyoung commented:
> Removing `a+b` is a breaking change and renders existing contracts incompatible. Shouldn't we first warn that `a+b` is deprecated and then remove it in a later release?

`a+b` for strings was always a non-standard extension. It is not used much; for example, in the [solana program examples](https://github.com/solana-developers/program-examples) it is not used (I've also just tested that).

As such I don't think there is need for deprecation and eventual removal. I've added an entry to `CHANGELOG.md`.

xermicus reviewed (approved):
LGTM, I just see some potential to optimize the concatenation into a no-op when one operand is an empty constant:

```solidity
function f(string a) public returns (string) {
    return string.concat(a, "");
}
```

```
block0: # entry
        ty:string %a = (arg #0)
        return (builtin Concat ((arg #0), (alloc string uint32 0 "")))
```

seanyoung commented:
> LGTM, I just see some potential to optimize the concatenation into a no-op when one operand is an empty constant:
> 
> ```solidity
> function f(string a) public returns (string) {
>     return string.concat(a, "");
> }
> ```
> 
> ```
> block0: # entry
>         ty:string %a = (arg #0)
>         return (builtin Concat ((arg #0), (alloc string uint32 0 "")))
> ```

Nice, good catch. I've fixed this and added a test case.

LucasSte commented:
> > Removing `a+b` is a breaking change and renders existing contracts incompatible. Shouldn't we first warn that `a+b` is deprecated and then remove it in a later release?
> 
> It is not used much; 

We don't know if it is not used much. There is no way to prove it. The solana examples is a small set of what people use Solang for. Polkadot also support this feature.

Instead of deprecating then, what do you think of suggesting people to use `string.concat` when erroring for `string + string`?

LucasSte commented on src/codegen/constant_folding.rs:
```diff
@@ -1924,51 +1952,89 @@ fn string_compare(
+                ..
+            }) = &mut last
+            {
+                if let Expression::NumberLiteral { value, .. } = size.as_mut() {
+                    *value += bs.len();
+                } else {
+                    unreachable!();
```
```suggestion
                let Expression::NumberLiteral { value, .. } = size.as_mut() else {
                    unreachable!();
                 }
                 *value += bs.len();
```

seanyoung commented:
> Instead of deprecating then, what do you think of suggesting people to use `string.concat` when erroring for `string + string`?

I've added a diagnostic for this. I think it's useful in general for folks when they may try to concatenate two strings.
